feugy/tabulous
# feat(web): introduces lobbies

### :book: What's in there?

Coordinating a group of people to meet and play together requires both async (find the appropriate time) and sync (decide which game to play) medium.
It's expected for players to use 3rd party (instant or mail) messaging systems, but Tabulous can easily facilitate this with waiting rooms, or lobbies.

Any player can create one (or several) lobbies, and invite other players, the same way they would with full games.
This would display a side panel with peer's avatars/videos and a discussion thread, on the home page, next to the game catalog.

Players can synchronously join and discuss, or asynchronously exchange on the game thread.
Later on, the lobby can be promoted to a full game, by any participant, by selecting any item in the catalog.

Are included here:

- fix(web): peers stay on the game page when owner deletes the game.
- feat(web): creates, deletes, join, promote and invite user to lobbies.
- feat(web): uses a browser notification when tab is inactive and being invited.
- refactor(web): use server game data when joining a game/lobby as peer, instead of waiting for the host to send it.
- refactor(web): moves some components from game route folder to components folder.
- refactor(web): removes page footer from layout component.

### :test_tube: How to test?

- on player A's home page, create a new lobby.
   > the invite dialogue appears.
- invite player B.
   > the side panel, with player B's avatar and discussion appears.
- leave a greeting message.
   > it appears in the discussion thread.
- click on the created lobby's invite button to invite player C.
   > player C's avatar appears in the side panel.
- in a different browser, connect as player B or C.
   > the lobby appears in the game list.
- click on the lobby.
   > the side panel appears, with player A's video, player B (or C)'s video, player C (or B)'s avatar, the discussion and the greeting message.
- chat a little bit, then select a game, to promote the lobby (any player).
   > you are both navigated to the game page, with meshes, or game parameters (depending on the game).
- the player who promoted the lobby into a full game goes back to home page.
   > a notification appear in the other player's game.
- remove the game.
   > the playing player is redirected back to their home page with a notification.

### :up: Notes to reviewers

Technically speaking, a lobby is a "kind-less" game. Because there is no kind, there is no descriptor, and therefore no meshes, hands no cameras.

Promoting a lobby is assigning a kind to it, so the server could apply the relevant descriptor. When promoting to game, all players and guest (including the owner) become player guest, so they could provide any required game parameters.

This leaves an unsolved issue: because there could be up to 8 persons in a lobby, some may be omitted from the final game. Players can't decide who, yet.
